syft/pkg/cataloger/generic/lookup.go

```diff
@@ -0,0 +1,85 @@
+package generic
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+	"text/template"
+
+	"github.com/anchore/syft/internal"
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/pkg/cataloger/internal/unionreader"
+	"github.com/anchore/syft/syft/source"
+)
+
+// Lookup is a generic package lookup that can be used to match a package definition
+// To a file that meets the given content criteria of the EvidencePatternTemplates.
+type Lookup struct {
```

Comment:
Would this be more of a `Binary` cataloger instead of `Lookup`? I might be misunderstanding the purpose. Ok, I think I understand now.

This seems like it's essentially doing what the catalogers are doing already, is this abstraction needed? It might be a little simpler -- since we're already adding a node binary cataloger -- to just make the `Find` function below a utility function for the `Package` and `EvidencePatternTemplates` and have it called by the node binary cataloger (which it already is doing) and specify the file path patterns in the `NewNodeBinaryCataloger`

Reply:
I think it's helpful to talk about some of the newer code that was recently merged to help explain the reason `Lookup` exists and why this still is a `BinaryCataloger`.

https://github.com/anchore/syft/blob/main/syft/pkg/cataloger/generic/cataloger.go <-- Here is the upgraded `Generic` cataloger that we're utilizing for `node-binary-cataloger` construction.

https://github.com/anchore/syft/blob/4fea2fa71e926acc27947f8b697706e9efdbc85d/syft/pkg/cataloger/javascript/cataloger.go#L40-L43

`parseNodeBinary` is the `Parser` that was added here which is associated with a location:
https://github.com/anchore/syft/blob/4fea2fa71e926acc27947f8b697706e9efdbc85d/syft/pkg/cataloger/generic/cataloger.go#L83-L92

In our case, the parser is going to look at every location that has a MIMEType in this list:
https://github.com/anchore/syft/blob/bd9007fc0e3de88685a55d3aafedd7af6e233922/internal/mimetype_helper.go#L55-L62

It is not matching on the glob patterns from the get-go.

Rather than write `nodeBinaryParser` `goBinaryParser` `cargoBinaryParser`, etc with the duplicated logic of further refinement via `FilepathPatterns` and then content analysis with`checkEvidenceInContent`, I wanted to condense that in a single place like the classification does for the `file` package ==> `Lookup` was born.

So this is not as a cataloger that's initially matching on `glob` in the first place, we should instead think of it as a cataloger that pulls in all files with the executable mime-type and then has a mechanism to further refine that executable list via glob/regex for a list of FilepathPatterns.

Of note! If FilepathPatterns is empty, we should just do a content analysis of `EvidencePatternTemplate` against all files returned by the cataloger. I'll make that update now!.

Here is an example of this where we are doing this with the new `GoBinary` parsing:
https://github.com/anchore/syft/blob/6826d7603bfcec9f3d2654609987ffb27f9eb7dd/syft/pkg/cataloger/golang/parse_go_binary.go#L41

So it's not `doing what the catalogers are doing already`, since we need some way to match against `EvidencePatternTemplates` across ecosystems that are not duplicated

The result of this generic `EvidencePatternTemplate` matching should then be returned to the parsers to assign package/ecosystem-specific metadata for the given lookup. The package in the parser is still mutable after coming back from the `Lookup`

`Lookup` also serves the purpose in giving a common way to configure the `Package Name` and potentially other domain-specific details you're searching for that will be constructed on a HIT against those pattern templates.

Reply:
Got it -- thank you for the detailed response here! So the way I see this is: we have a multi-factor match we're looking for: _executable_ AND _matching the node regex_, essentially a new cataloger builder function that would be something like `WithParserByMimeTypesAndPattern`, following the current pattern (I won't get into the naming or combinatorial explosion here, which could both be valid)? I also don't have _any_ problem with a configuration struct, as you noted, or whichever direction (utility function or method) it's used for the `Locator`. My main question is around conflating file lookups in the midst of it all, which seem to be a `Cataloger` concern.

Reply:
We don't have to call it lookup then. `Matcher` `Classifier` `Identifier` <--? pick you're favorite

Reply:
It's not the naming as much, but the fact it's doing file matching, which the `Cataloger` is already doing and could be expanded upon to do both _executable_ and _pattern_ matches together. In this case I'd probably rename it from `Locator` as well, to something like `ContentMatcher`